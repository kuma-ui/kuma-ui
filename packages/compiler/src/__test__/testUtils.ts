import { compile } from "../compile";
import { compileSync } from "..";

type Result = ReturnType<typeof compile>;

export function getExpectSnapshot(result: Result) {
  return `
  ${result.css}
  
  ${result.code}
  `;
}

type CompileResult = ReturnType<typeof compileSync>;

export function getExpectSnapshotSync(result: CompileResult) {
  return `
  ${result?.css}
  
  ${result?.code}
  `;
}
