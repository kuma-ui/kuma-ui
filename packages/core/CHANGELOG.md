# @kuma-ui/core

## 1.5.6

### Patch Changes

- [#398](https://github.com/kuma-ui/kuma-ui/pull/398) [`08632b1`](https://github.com/kuma-ui/kuma-ui/commit/08632b17471b62da3d2e3c0d6fec0097ee1cd514) Thanks [@yukukotani](https://github.com/yukukotani)! - Fix `DOMException: Failed to execute 'deleteRule' on 'CSSStyleSheet': The index provided (4294967295) is larger than the maximum index (0).`

- Updated dependencies [[`b3490c1`](https://github.com/kuma-ui/kuma-ui/commit/b3490c10c30c9fa65167bb51554903397f4a90a5), [`a0dc9cd`](https://github.com/kuma-ui/kuma-ui/commit/a0dc9cda8ecfad419be8053c1237d4d6551ef77b)]:
  - @kuma-ui/system@1.7.3

## 1.5.5

### Patch Changes

- [#385](https://github.com/kuma-ui/kuma-ui/pull/385) [`407a55e`](https://github.com/kuma-ui/kuma-ui/commit/407a55e4b2e4f814fd4a3ea3539044bb2d5d0937) Thanks [@kotarella1110](https://github.com/kotarella1110)! - refactor: improve toCssUnit logic to reduce bundle size

- Updated dependencies [[`407a55e`](https://github.com/kuma-ui/kuma-ui/commit/407a55e4b2e4f814fd4a3ea3539044bb2d5d0937), [`2bb1f6d`](https://github.com/kuma-ui/kuma-ui/commit/2bb1f6d449f87d5fab6a1fd23cd1a2f925478c7d)]:
  - @kuma-ui/system@1.7.2

## 1.5.4

### Patch Changes

- [#380](https://github.com/kuma-ui/kuma-ui/pull/380) [`d2fe302`](https://github.com/kuma-ui/kuma-ui/commit/d2fe302a0d841c19cd628a94e12423a89599f3cc) Thanks [@yukukotani](https://github.com/yukukotani)! - avoid false-positive hasDynamicProps

- Updated dependencies [[`afd4d57`](https://github.com/kuma-ui/kuma-ui/commit/afd4d579af008fcd9fd8f602956ff665590ffca5)]:
  - @kuma-ui/system@1.7.1

## 1.5.3

### Patch Changes

- [#372](https://github.com/kuma-ui/kuma-ui/pull/372) [`3840e83`](https://github.com/kuma-ui/kuma-ui/commit/3840e835ed3c3c6bb707a565ecec91f2e737eb18) Thanks [@AdePhil](https://github.com/AdePhil)! - chore: Refactor ts-pattern to use switch/if-else

- Updated dependencies [[`d439b72`](https://github.com/kuma-ui/kuma-ui/commit/d439b72e9dd651ccd65a889c1dce5b3fc4e54a11), [`aa23d92`](https://github.com/kuma-ui/kuma-ui/commit/aa23d92f93b0373b91c7b0d431a5745aa06380b9)]:
  - @kuma-ui/system@1.7.0

## 1.5.2

### Patch Changes

- Updated dependencies [[`b5415ed`](https://github.com/kuma-ui/kuma-ui/commit/b5415ed684c5212a4dd6d6e10d43969164375233)]:
  - @kuma-ui/system@1.6.0

## 1.5.1

### Patch Changes

- [#311](https://github.com/kuma-ui/kuma-ui/pull/311) [`69b90fa`](https://github.com/kuma-ui/kuma-ui/commit/69b90fadedd338d552aa6c128226ea7732f7fafa) Thanks [@poteboy](https://github.com/poteboy)! - Update prettier from 2.8 to 3.0.3

- [#351](https://github.com/kuma-ui/kuma-ui/pull/351) [`e8e5195`](https://github.com/kuma-ui/kuma-ui/commit/e8e5195706f90e99a0236d76c5825e5060425b7a) Thanks [@yukukotani](https://github.com/yukukotani)! - Override original prop types with `as` component props

- Updated dependencies [[`d8b82dc`](https://github.com/kuma-ui/kuma-ui/commit/d8b82dc7e851afa994eda664e09acff41d04ba6c)]:
  - @kuma-ui/system@1.5.2

## 1.5.0

### Minor Changes

- [#343](https://github.com/kuma-ui/kuma-ui/pull/343) [`1e31d15`](https://github.com/kuma-ui/kuma-ui/commit/1e31d159d968e87054bb64155be1618d568be172) Thanks [@poteboy](https://github.com/poteboy)! - Implement ForwardRef in Components

- [#342](https://github.com/kuma-ui/kuma-ui/pull/342) [`2c3da38`](https://github.com/kuma-ui/kuma-ui/commit/2c3da38efccea069e9a4ac8e4fa7f5e7d50d0db6) Thanks [@mmkal](https://github.com/mmkal)! - styled-components syntax parity

## 1.4.1

### Patch Changes

- Updated dependencies [[`6bfcab7`](https://github.com/kuma-ui/kuma-ui/commit/6bfcab7b85de6ff020dec3e65cf936aebd14be54)]:
  - @kuma-ui/system@1.5.1

## 1.4.0

### Minor Changes

- [#240](https://github.com/kuma-ui/kuma-ui/pull/240) [`4073398`](https://github.com/kuma-ui/kuma-ui/commit/4073398d58e011f0afb351025e9b326ff56c0e36) Thanks [@poteboy](https://github.com/poteboy)! - Enhance Performance by Optimizing Compiler for Static HTML Conversion

- [#321](https://github.com/kuma-ui/kuma-ui/pull/321) [`8e1a211`](https://github.com/kuma-ui/kuma-ui/commit/8e1a211860281f2fb7eb94bb2f8680300ebf7b7c) Thanks [@yukukotani](https://github.com/yukukotani)! - Introduce `defaultProps` in component theme

### Patch Changes

- [#304](https://github.com/kuma-ui/kuma-ui/pull/304) [`1b7982e`](https://github.com/kuma-ui/kuma-ui/commit/1b7982e5d8b9ab4f862d1ff113353380b411129e) Thanks [@yukukotani](https://github.com/yukukotani)! - Improve autocompletion experience of `baseStyle` and `variants`

- [#324](https://github.com/kuma-ui/kuma-ui/pull/324) [`0fc8f4d`](https://github.com/kuma-ui/kuma-ui/commit/0fc8f4d86972517052aacb982fb9c9d94788353a) Thanks [@kotarella1110](https://github.com/kotarella1110)! - improve createTheme type

- [#316](https://github.com/kuma-ui/kuma-ui/pull/316) [`f441c1d`](https://github.com/kuma-ui/kuma-ui/commit/f441c1d0b3a5e2e13d83aeff4984ad64702b69d9) Thanks [@yukukotani](https://github.com/yukukotani)! - Exclude vitest.setup.ts from dist files

- Updated dependencies [[`8d68550`](https://github.com/kuma-ui/kuma-ui/commit/8d685505e17beceafd43d9d3618342e1246032e4), [`031bc85`](https://github.com/kuma-ui/kuma-ui/commit/031bc851c140d5fbbd7a6b40116c2b894113940e), [`8e1a211`](https://github.com/kuma-ui/kuma-ui/commit/8e1a211860281f2fb7eb94bb2f8680300ebf7b7c)]:
  - @kuma-ui/system@1.5.0
  - @kuma-ui/sheet@1.3.0

## 1.3.0

### Minor Changes

- [#293](https://github.com/kuma-ui/kuma-ui/pull/293) [`638d61d`](https://github.com/kuma-ui/kuma-ui/commit/638d61df0a6dd355d8bb4847e5752877a9895d71) Thanks [@yukukotani](https://github.com/yukukotani)! - Add support for z-index token

### Patch Changes

- Updated dependencies [[`638d61d`](https://github.com/kuma-ui/kuma-ui/commit/638d61df0a6dd355d8bb4847e5752877a9895d71), [`3414300`](https://github.com/kuma-ui/kuma-ui/commit/341430056cff0e4097e0b066726e5c110219eac9), [`314096c`](https://github.com/kuma-ui/kuma-ui/commit/314096cde1d45ffa35c93ef49bddacb7e745606b)]:
  - @kuma-ui/system@1.4.0
  - @kuma-ui/sheet@1.2.0

## 1.2.0

### Minor Changes

- [#274](https://github.com/kuma-ui/kuma-ui/pull/274) [`bc1a7f9`](https://github.com/kuma-ui/kuma-ui/commit/bc1a7f979f5539a4e2d1d2785dc2a893eee076eb) Thanks [@yukukotani](https://github.com/yukukotani)! - Add support for spacing theme token

- [#289](https://github.com/kuma-ui/kuma-ui/pull/289) [`10f9811`](https://github.com/kuma-ui/kuma-ui/commit/10f98111d542f1fa45a6cda18ac872bec7b40910) Thanks [@yukukotani](https://github.com/yukukotani)! - Add support for size token

- [#291](https://github.com/kuma-ui/kuma-ui/pull/291) [`066d043`](https://github.com/kuma-ui/kuma-ui/commit/066d04383d3f4e19fb0460b184e8aa9ed67ae41d) Thanks [@yukukotani](https://github.com/yukukotani)! - Add support for radius token

### Patch Changes

- [#286](https://github.com/kuma-ui/kuma-ui/pull/286) [`144e1d0`](https://github.com/kuma-ui/kuma-ui/commit/144e1d01a852ca21fc4a89da7f5da2a1f0a4f955) Thanks [@MH4GF](https://github.com/MH4GF)! - chore: modified existing test cases with the introduction of testing-library/jest-dom

- Updated dependencies [[`bc1a7f9`](https://github.com/kuma-ui/kuma-ui/commit/bc1a7f979f5539a4e2d1d2785dc2a893eee076eb), [`10f9811`](https://github.com/kuma-ui/kuma-ui/commit/10f98111d542f1fa45a6cda18ac872bec7b40910), [`066d043`](https://github.com/kuma-ui/kuma-ui/commit/066d04383d3f4e19fb0460b184e8aa9ed67ae41d)]:
  - @kuma-ui/system@1.3.0
  - @kuma-ui/sheet@1.1.0

## 1.1.3

### Patch Changes

- [#276](https://github.com/kuma-ui/kuma-ui/pull/276) [`e276aea`](https://github.com/kuma-ui/kuma-ui/commit/e276aea4dbd933800de5237d0411df9097451dfe) Thanks [@naruaway](https://github.com/naruaway)! - fix: Replace broken generateKey with fast-json-stable-stringify for pseudo props

- [#278](https://github.com/kuma-ui/kuma-ui/pull/278) [`a03add5`](https://github.com/kuma-ui/kuma-ui/commit/a03add5b2a0b4ec9bcd6d2fd636c6d03a14f1636) Thanks [@MH4GF](https://github.com/MH4GF)! - maintenance: setup testing-library/jest-dom and add component test for Box

- [#272](https://github.com/kuma-ui/kuma-ui/pull/272) [`2329a1d`](https://github.com/kuma-ui/kuma-ui/commit/2329a1d9ea2fe4cf2d61cdf4ea87983c75b0d8fb) Thanks [@taishinaritomi](https://github.com/taishinaritomi)! - feat: support font theme tokens

- Updated dependencies [[`2329a1d`](https://github.com/kuma-ui/kuma-ui/commit/2329a1d9ea2fe4cf2d61cdf4ea87983c75b0d8fb)]:
  - @kuma-ui/sheet@1.0.2
  - @kuma-ui/system@1.2.1

## 1.1.2

### Patch Changes

- Updated dependencies [[`022fdec`](https://github.com/kuma-ui/kuma-ui/commit/022fdecfd1c2494974e9965ddca31aebf03d1dc2), [`90a9e65`](https://github.com/kuma-ui/kuma-ui/commit/90a9e6504d9e11f164f2373a8968e8eef7130b9d)]:
  - @kuma-ui/system@1.2.0

## 1.1.1

### Patch Changes

- Updated dependencies [[`8a42a22`](https://github.com/kuma-ui/kuma-ui/commit/8a42a225359bad6d2bf35a3c63d897ae02f59392)]:
  - @kuma-ui/system@1.1.0

## 1.1.0

### Minor Changes

- [#224](https://github.com/poteboy/kuma-ui/pull/224) [`b9110d7`](https://github.com/poteboy/kuma-ui/commit/b9110d7e17c7ad16b70604f2272adb53c96c5544) Thanks [@rabitarochan](https://github.com/rabitarochan)! - fix: resolving paths on Windows

## 1.0.1

### Patch Changes

- Updated dependencies [[`ee93574`](https://github.com/poteboy/kuma-ui/commit/ee9357493694118f6c9c829c9107221688a65ab8)]:
  - @kuma-ui/sheet@1.0.1
  - @kuma-ui/system@1.0.1

## 1.0.0

### Major Changes

- [`8ada2cd`](https://github.com/poteboy/kuma-ui/commit/8ada2cd64f144103cf611fc0990f4f74bcc19969) Thanks [@poteboy](https://github.com/poteboy)! - üêª‚Äç‚ùÑÔ∏è

### Patch Changes

- Updated dependencies [[`8ada2cd`](https://github.com/poteboy/kuma-ui/commit/8ada2cd64f144103cf611fc0990f4f74bcc19969)]:
  - @kuma-ui/sheet@1.0.0
  - @kuma-ui/system@1.0.0

## 0.6.0-next.1

### Minor Changes

- [#126](https://github.com/poteboy/kuma-ui/pull/126) [`c4b7318`](https://github.com/poteboy/kuma-ui/commit/c4b73184ca8e9c659734f9547143806a3a7d222a) Thanks [@poteboy](https://github.com/poteboy)! - Introducing Theming, Hybrid CSS-in-JS, and Universal UI Components

### Patch Changes

- Updated dependencies [[`c4b7318`](https://github.com/poteboy/kuma-ui/commit/c4b73184ca8e9c659734f9547143806a3a7d222a)]:
  - @kuma-ui/system@0.7.0-next.1
  - @kuma-ui/sheet@0.6.0-next.1

## 0.6.0-next.0

### Minor Changes

- [#126](https://github.com/poteboy/kuma-ui/pull/126) [`c89e97c`](https://github.com/poteboy/kuma-ui/commit/c89e97c2f2d3fc56586d0e02d5d8ff873b9eee6a) Thanks [@poteboy](https://github.com/poteboy)! - Introducing Theming, Hybrid CSS-in-JS, and Universal UI Components in @kuma-ui/core

### Patch Changes

- Updated dependencies [[`c89e97c`](https://github.com/poteboy/kuma-ui/commit/c89e97c2f2d3fc56586d0e02d5d8ff873b9eee6a)]:
  - @kuma-ui/system@0.7.0-next.0
  - @kuma-ui/sheet@0.6.0-next.0

## 0.5.0

### Minor Changes

- [#139](https://github.com/poteboy/kuma-ui/pull/139) [`6d0ba6d`](https://github.com/poteboy/kuma-ui/commit/6d0ba6db946f38de2bb3ea51b288daddfbb67196) Thanks [@poteboy](https://github.com/poteboy)! - Pre-release Bug Fix: Limit style props extraction to Kuma UI components

## 0.4.2

### Patch Changes

- 8af4f8d: chore: Automate synchronization of README.md across packages.
- Updated dependencies [8af4f8d]
  - @kuma-ui/system@0.6.2

## 0.4.1

### Patch Changes

- 558823e: fix classname hash
  - @kuma-ui/system@0.6.1

## 0.4.0

### Minor Changes

- Fully support RSC & Fully support pseudo elements & classes & throw an error when accessing any property of 'k' at runtime

### Patch Changes

- Updated dependencies
  - @kuma-ui/system@0.6.0

## 0.3.1

### Patch Changes

- @kuma-ui/system@0.5.1

## 0.3.0

### Minor Changes

- Documentation site

### Patch Changes

- Updated dependencies
  - @kuma-ui/system@0.5.0

## 0.2.0

### Minor Changes

- Added js-doc to css props type & Add css Function

## 0.1.0

### Minor Changes

- experimental support for Next.js 13.4 App Router and React Server Components

## 0.0.4

### Patch Changes

- Overriding styles defined within tagged template literals using style props

## 0.0.3

### Patch Changes

- Replace global whitespace removal regex with more specific regex for CSS properties #23
  Refactor HTML Tags. #24
  Fix incorrect setup example for Next.js #25
  fix position 'bottom, inset' css prop values #26

## 0.0.2

### Patch Changes

- initial release of kuma-ui packages
