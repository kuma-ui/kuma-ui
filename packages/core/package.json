{
  "name": "@kuma-ui/core",
  "version": "1.5.6",
  "description": "🐻 Kuma UI is a utility-first, zero-runtime CSS-in-JS library that offers an outstanding developer experience and optimized performance.",
  "repository": {
    "type": "git",
    "url": "git+https://github.com/kuma-ui/kuma-ui.git"
  },
  "keywords": [
    "react",
    "css-in-js",
    "zero-runtime",
    "front-end",
    "design-system",
    "styled-components"
  ],
  "homepage": "https://www.kuma-ui.com",
  "main": "dist/index.js",
  "module": "dist/index.mjs",
  "types": "dist/index.d.ts",
  "sideEffects": false,
  "exports": {
    ".": {
      "types": "./dist/index.d.ts",
      "require": "./dist/index.js",
      "import": "./dist/index.mjs"
    },
    "./components/componentList": {
      "types": "./dist/components/componentList.d.ts",
      "require": "./dist/components/componentList.js",
      "import": "./dist/components/componentList.mjs"
    }
  },
  "typesVersions": {
    "*": {
      "components/componentList": [
        "./dist/components/componentList.d.ts"
      ]
    }
  },
  "files": [
    "dist"
  ],
  "scripts": {
    "build": "tsup",
    "test": "vitest run && vitest typecheck --run",
    "typecheck": "tsc --noEmit --composite false",
    "lint": "eslint './src/**/*.{js,ts,jsx,tsx}' --report-unused-disable-directives --max-warnings 0",
    "lint:fix": "eslint --fix './src/**/*.{js,ts,jsx,tsx}' --report-unused-disable-directives --max-warnings 0"
  },
  "dependencies": {
    "@kuma-ui/sheet": "workspace:*",
    "@kuma-ui/system": "workspace:*",
    "@types/stylis": "^4.2.0",
    "stylis": "^4.2.0"
  },
  "devDependencies": {
    "@testing-library/jest-dom": "^5.16.5",
    "@testing-library/react": "^14.0.0",
    "@types/react": "^18.0.32",
    "@types/testing-library__jest-dom": "^5.14.9",
    "@vitest/coverage-c8": "^0.31.4",
    "globby": "^13.2.0",
    "jsdom": "^22.1.0",
    "ts-node": "^10.9.1",
    "vitest": "^0.31.4"
  },
  "peerDependencies": {
    "@types/react": "^18.0.32",
    "next": ">=13.4.5",
    "react": ">=18.2.0"
  },
  "peerDependenciesMeta": {
    "@types/react": {
      "optional": true
    },
    "next": {
      "optional": true
    }
  },
  "publishConfig": {
    "access": "public"
  }
}
