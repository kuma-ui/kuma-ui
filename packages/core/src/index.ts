export { styled } from "./styled";
export { k } from "./k";
export { css, cx } from "./css";
export { createTheme, type Theme, type ThemeSystem } from "./theme";
export * from "./components";
export * from "./registry";
