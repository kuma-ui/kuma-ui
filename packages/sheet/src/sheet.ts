import { generateHash } from "./hash";
import { cssPropertyRegex, removeSpacesExceptInPropertiesRegex } from "./regex";
import { compile, serialize, stringify } from "stylis";

// to avoid cyclic dependency, we declare an exact same type declared in @kuma-ui/system
type ResponsiveStyle = {
  base: string;
  media: {
    [breakpoint: string]: string;
  };
};

type SystemStyle = {
  base: ResponsiveStyle["base"];
  responsive: ResponsiveStyle["media"];
  pseudo: {
    key: string;
    base: ResponsiveStyle["base"];
    responsive: ResponsiveStyle["media"];
  }[];
};

export { SystemStyle };

export class Sheet {
  private static instance: Sheet;
  private base: string[];
  private responsive: string[];
  private pseudo: string[];

  private css: string[];

  private constructor() {
    this.base = [];
    this.responsive = [];
    this.pseudo = [];
    this.css = [];
  }

  static getInstance() {
    if (!Sheet.instance) {
      Sheet.instance = new Sheet();
    }
    return Sheet.instance;
  }

  addRule(style: SystemStyle) {
    const className = "kuma-" + generateHash(JSON.stringify(style));
    this._addeBaseRule(className, style.base);
    for (const [breakpoint, css] of Object.entries(style.responsive)) {
      this._addMediaRule(className, css, breakpoint);
    }
    for (const [_, pseudo] of Object.entries(style.pseudo)) {
      this._addPseudoRule(className, pseudo);
    }
    return className;
  }

  private _addeBaseRule(className: string, css: string) {
    css = css.replace(cssPropertyRegex, "");
    this.base.push(`.${className}{${css}}`);
  }

  private _addMediaRule(
    className: string,
    css: string,
    breakpoint: string
  ): void {
    css = css.replace(cssPropertyRegex, "");
    const mediaCss =
      `@media (min-width: ${breakpoint}) { .${className} { ${css} } }`.replace(
        removeSpacesExceptInPropertiesRegex,
        ""
      );
    this.responsive.push(mediaCss);
  }

  private _addPseudoRule(
    className: string,
    pseudo: SystemStyle["pseudo"][number]
  ) {
    const css = pseudo.base.replace(cssPropertyRegex, "");
    const pseudoCss = `.${className}${pseudo.key} { ${css} }`.replace(
      removeSpacesExceptInPropertiesRegex,
      ""
    );
    this.pseudo.push(pseudoCss);
    for (const [breakpoint, _css] of Object.entries(pseudo.responsive)) {
      this._addMediaRule(`${className}${pseudo.key}`, _css, breakpoint);
    }
  }

  /**
   * parseCSS takes in raw CSS and parses it to valid CSS using Stylis.
   * It's useful for handling complex CSS such as media queries and pseudo selectors.
   */
  parseCSS(style: string): string {
    const id = "kuma-" + generateHash(style);
    const css = serialize(compile(`.${id}{${style}}`), stringify);
    this.css.push(css);
    return id;
  }

  removeDuplicates() {
    this.base = [...new Set(this.base)];
    this.responsive = [...new Set(this.responsive)];
    this.pseudo = [...new Set(this.pseudo)];
    this.css = [...new Set(this.css)];
  }

  getCSS(): string {
    this.removeDuplicates();
    return (
      this.base.join("") +
      this.responsive.join("") +
      this.pseudo.join("") +
      this.css.join("")
    );
  }

  reset() {
    this.base = [];
    this.responsive = [];
    this.pseudo = [];
    this.css = [];
  }
}

/**
 * The Sheet singleton class is responsible for managing the CSS rules, responsive
 * styles, and pseudo-classes used by Kuma-UI. It provides methods to add and retrieve
 * rules, and generates the final CSS output.
 */
export const sheet = Sheet.getInstance();
