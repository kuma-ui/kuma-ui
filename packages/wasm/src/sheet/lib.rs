/**
 * The generateHash function is a utility function that generates a hash value for a given input string using the MurmurHash algorithm.
 */
pub fn generate_hash(str: &str) -> String {
    static M: u32 = 0x5bd1e995;
    static R: u32 = 24;
    static SEED: u32 = 0x12345678;
    let len = str.len();

    let mut h = SEED ^ len as u32;

    for i in 0..len {
        let k = str.as_bytes()[i] as u32;
        let mut k = k.wrapping_mul(M);
        k ^= k >> R;
        k = k.wrapping_mul(M);

        h = h.wrapping_mul(M);
        h ^= k;
    }

    h ^= h >> 13;
    h = h.wrapping_mul(M);
    h ^= h >> 15;
    h.to_string()
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_generate_hash() {
        let input = "Hello, World!";
        let input2 = "Hello, World!";
        assert_eq!(generate_hash(input), generate_hash(input2));
        let input3 = "Hello, World";
        assert_ne!(generate_hash(input), generate_hash(input3));
    }
}
