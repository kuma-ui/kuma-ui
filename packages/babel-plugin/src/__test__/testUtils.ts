import { BabelFileResult, transformSync } from "@babel/core";
import plugin from "../";

export function babelTransform(
  code: string,
  runtime?: "classic" | "automatic"
): BabelFileResult {
  const filename = "test.tsx";
  const result1 = transformSync(code, {
    plugins: [plugin],
    filename,
  });

  if (result1 === null || result1.code == null)
    throw new Error(`Could not transform`);

  if (!runtime) {
    return result1;
  }

  const result2 = transformSync(result1.code, {
    presets: [
      "@babel/preset-typescript",
      [
        "@babel/preset-react",
        {
          runtime,
        },
      ],
    ],
    filename,
  });

  if (result2 === null || result2.code == null)
    throw new Error(`Could not transform`);

  return {
    metadata: result1.metadata,
    code: result2.code,
  };
}

export function getExpectSnapshot(result: BabelFileResult) {
  return `
${(result.metadata as { css: string } | undefined)?.css}

${result.code}
`;
}
