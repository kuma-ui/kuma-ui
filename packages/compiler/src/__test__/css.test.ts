import { describe, beforeEach, test, expect } from "vitest";
import { compileSync } from "..";
import { getExpectSnapshotSync } from "./testUtils";
import { theme, sheet } from "@kuma-ui/sheet";

describe("css", () => {
  beforeEach(() => {
    sheet.reset();
    theme.setUserTheme({
      colors: {
        "colors.primary": "red",
      },
    });
  });

  describe.each([{ wasm: true }, { wasm: false }])(
    "Snapshot tests with wasm: %p",
    ({ wasm }) => {
      test("basic usage should match snapshot", () => {
        // Arrange
        const inputCode = `
              import { css } from '@kuma-ui/core'
              const style = css\`color: red;\`
            `;
        // Act
        const result = compileSync({ code: inputCode, id: "test.tsx", wasm });
        // Assert
        expect(getExpectSnapshotSync(result)).toMatchSnapshot();
      });

      test("using space props should match snapshot", () => {
        // Arrange
        const inputCode = `
              import { css } from '@kuma-ui/core'
              const style = css\`padding: 2px;\`
            `;
        // Act
        const result = compileSync({ code: inputCode, id: "test.tsx", wasm });
        // Assert
        expect(getExpectSnapshotSync(result)).toMatchSnapshot();
      });

      test("using pseudo props should match snapshot", () => {
        // Arrange
        const inputCode = `
              import { css } from '@kuma-ui/core'
              const style = css\`
              &:hover {
                color: red;
              }
              \`
            `;
        // Act
        const result = compileSync({ code: inputCode, id: "test.tsx", wasm });
        // Assert
        expect(getExpectSnapshotSync(result)).toMatchSnapshot();
      });

      test("placeholder usage should match snapshot", () => {
        // Arrange
        const originalCode = `
              import { css } from '@kuma-ui/core'
              const style = css\`color: red;\`
            `;
        const inputCode = `
              import { css } from '@kuma-ui/core'
              const style = css\`color: t("colors.primary");\`
            `;
        // Act
        const result = compileSync({ code: inputCode, id: "test.tsx", wasm });
        const originalResult = compileSync({
          code: originalCode,
          id: "test.tsx",
          wasm,
        });
        // Assert
        expect(getExpectSnapshotSync(result)).toEqual(
          getExpectSnapshotSync(originalResult),
        );
      });
    },
  );
});
