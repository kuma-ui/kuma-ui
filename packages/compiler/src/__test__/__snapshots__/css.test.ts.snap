// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`css > Snapshot tests with wasm: %p { wasm: false } > basic usage should match snapshot 1`] = `
"
  .🐻-136547911{color:red;}
  
  import { Box as __Box } from \\"@kuma-ui/core\\";
import __KUMA_REACT__ from \\"react\\";
import { css } from '@kuma-ui/core';
const style = \\"🐻-136547911\\";
  "
`;

exports[`css > Snapshot tests with wasm: %p { wasm: false } > using pseudo props should match snapshot 1`] = `
"
  .🐻-2051565201:hover{color:red;}
  
  import { Box as __Box } from \\"@kuma-ui/core\\";
import __KUMA_REACT__ from \\"react\\";
import { css } from '@kuma-ui/core';
const style = \\"🐻-2051565201\\";
  "
`;

exports[`css > Snapshot tests with wasm: %p { wasm: false } > using space props should match snapshot 1`] = `
"
  .🐻-620795649{padding:2px;}
  
  import { Box as __Box } from \\"@kuma-ui/core\\";
import __KUMA_REACT__ from \\"react\\";
import { css } from '@kuma-ui/core';
const style = \\"🐻-620795649\\";
  "
`;

exports[`css > Snapshot tests with wasm: %p { wasm: true } > basic usage should match snapshot 1`] = `
"
  .🐻-136547911{color:red;}
  
  import {Box as __Box} from '@kuma-ui/core';import __KUMA_REACT__ from 'react';import {css} from '@kuma-ui/core';const style=\\"🐻-136547911\\";
  "
`;

exports[`css > Snapshot tests with wasm: %p { wasm: true } > using pseudo props should match snapshot 1`] = `
"
  .🐻-2051565201:hover{color:red;}
  
  import {Box as __Box} from '@kuma-ui/core';import __KUMA_REACT__ from 'react';import {css} from '@kuma-ui/core';const style=\\"🐻-2051565201\\";
  "
`;

exports[`css > Snapshot tests with wasm: %p { wasm: true } > using space props should match snapshot 1`] = `
"
  .🐻-620795649{padding:2px;}
  
  import {Box as __Box} from '@kuma-ui/core';import __KUMA_REACT__ from 'react';import {css} from '@kuma-ui/core';const style=\\"🐻-620795649\\";
  "
`;
